File: experiments/000-PipelineChecks/cells/python-pytest_1-Empty_free/2024-1c-parcial1/20260921-202040-52205/output/project/src/simshotels/floor.py
from simshotels.room import Room


class Floor:

    def __init__(self, a_number_of_rooms=None, a_price_list=None):
        self._rooms = []
        self._prices = {}
        if a_number_of_rooms is not None and a_price_list is not None:
            self.set_number_of_rooms(a_number_of_rooms)
            self.set_prices(a_price_list)

    # testing

    def is_available(self):
        return self.total_rooms() == self.total_rooms_available()

    # accessing

    def rooms(self):
        return self._rooms

    # room operations

    def receive_at_room(self, a_guest_type, a_room_number):
        if a_room_number < 1 or a_room_number > len(self._rooms):
            Floor.signal_room_number_does_not_exist()
        room = self._rooms[a_room_number - 1]
        room.receive(a_guest_type)

    def receive_with_reservation_at_room(self, a_guest_type, a_room_number):
        if a_room_number < 1 or a_room_number > len(self._rooms):
            Floor.signal_room_number_does_not_exist()
        room = self._rooms[a_room_number - 1]
        room.receive_with_reservation(a_guest_type)

    def reserve_room(self, a_room_number):
        if a_room_number < 1 or a_room_number > len(self._rooms):
            Floor.signal_room_number_does_not_exist()
        room = self._rooms[a_room_number - 1]
        room.reserve()

    # accounting

    def total_losses(self):
        return sum(room.loss_using_if_absent_guest_type(self._prices, lambda: Floor.signal_unknown_guest_type()) for room in self._rooms)

    def total_profits(self):
        return sum(room.profit_using_if_absent_guest_type(self._prices, lambda: Floor.signal_unknown_guest_type()) for room in self._rooms if room.is_occupied())

    # setters

    def set_number_of_rooms(self, a_number_of_rooms):
        Floor.assert_is_positive_if_false(a_number_of_rooms, lambda: Floor.signal_number_of_rooms_must_be_positive())
        Floor.assert_is_integer_if_false(a_number_of_rooms, lambda: Floor.signal_number_of_rooms_must_be_integer())

        self.initialize_rooms_with(a_number_of_rooms)

    def set_prices(self, a_price_list):
        if len(a_price_list) == 0:
            Floor.signal_no_prices()

        for price in a_price_list.values():
            Floor.assert_is_positive_if_false(price, lambda: Floor.signal_price_must_be_positive())

        self.initialize_with(a_price_list)

    # totals

    def total_rooms(self):
        return len(self._rooms)

    def total_rooms_available(self):
        return sum(1 for room in self._rooms if room.is_available())

    def total_rooms_occupied(self):
        return sum(1 for room in self._rooms if room.is_occupied())

    def total_rooms_reserved(self):
        return sum(1 for room in self._rooms if room.is_reserved())

    # initialization

    def initialize_rooms_with(self, a_number_of_rooms):
        self._rooms = []

        ix = 1
        while ix <= a_number_of_rooms:
            self._rooms.append(Room())
            ix = ix + 1

    def initialize_with(self, a_price_list):
        self._prices = a_price_list

    # assertions (class side)

    @staticmethod
    def assert_is_integer_if_false(a_number, signals_block):
        if not (isinstance(a_number, int) and not isinstance(a_number, bool)):
            signals_block()

    @staticmethod
    def assert_is_positive_if_false(a_number, signals_block):
        if not (a_number > 0):
            signals_block()

    # error messages (class side)

    @staticmethod
    def number_of_rooms_must_be_integer_error_description():
        return "Number of rooms must be integer"

    @staticmethod
    def number_of_rooms_must_be_positive_error_description():
        return "The number of rooms must be positive"

    @staticmethod
    def price_must_be_integer_error_description():
        return "Price must be integer"

    @staticmethod
    def price_must_be_positive_error_description():
        return "Price must be positive"

    @staticmethod
    def prices_list_cannot_be_empty_error_description():
        return "Prices List must not be empty"

    @staticmethod
    def room_number_does_not_exist_error_description():
        return "Room number does not exist"

    @staticmethod
    def unknown_guest_type_error_description():
        return "The guest type is unknown"

    # exceptions (class side)

    @staticmethod
    def signal_no_prices():
        raise RuntimeError(Floor.prices_list_cannot_be_empty_error_description())

    @staticmethod
    def signal_number_of_rooms_must_be_integer():
        raise RuntimeError(Floor.number_of_rooms_must_be_integer_error_description())

    @staticmethod
    def signal_number_of_rooms_must_be_positive():
        raise RuntimeError(Floor.number_of_rooms_must_be_positive_error_description())

    @staticmethod
    def signal_price_must_be_integer():
        raise RuntimeError(Floor.price_must_be_integer_error_description())

    @staticmethod
    def signal_price_must_be_positive():
        raise RuntimeError(Floor.price_must_be_positive_error_description())

    @staticmethod
    def signal_room_number_does_not_exist():
        raise RuntimeError(Floor.room_number_does_not_exist_error_description())

    @staticmethod
    def signal_unknown_guest_type():
        raise RuntimeError(Floor.unknown_guest_type_error_description())

File: experiments/020-PythonVsJava/cells/java-gradle_1-Empty_free/2024-1c-parcial1/20260921-203652-64546/analysis/acceptance-project/src/main/java/simshotels/Floor.java
package simshotels;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Predicate;
import java.util.function.Supplier;
import java.util.function.ToIntFunction;

public class Floor {

    private final List<Room> rooms;
    private final Map<String, Integer> prices;

    // initialization

    private Floor(int aNumberOfRooms, Map<String, Integer> aPriceList) {
        prices = new HashMap<>(aPriceList);
        rooms = new ArrayList<>();

        for (int roomNumber = 1; roomNumber <= aNumberOfRooms; roomNumber++) {
            rooms.add(new Room());
        }
    }

    // instance creation (class side)

    public static Floor withNumberOfRoomsAndPrices(Number aNumberOfRooms, Map<String, Integer> aPriceList) {
        Floor.assertIsPositiveIfFalse(aNumberOfRooms, () -> Floor.signalNumberOfRoomsMustBePositive());
        Floor.assertIsIntegerIfFalse(aNumberOfRooms, () -> Floor.signalNumberOfRoomsMustBeInteger());
        Floor.assertArePrices(aPriceList);

        return new Floor(aNumberOfRooms.intValue(), aPriceList);
    }

    // testing

    public boolean isAvailable() {
        return totalRooms() == totalRoomsAvailable();
    }

    // guests

    public void receiveAtRoom(String aGuestType, int aRoomNumber) {
        roomAt(aRoomNumber).receive(aGuestType);
    }

    public void receiveWithReservationAtRoom(String aGuestType, int aRoomNumber) {
        roomAt(aRoomNumber).receiveWithReservation(aGuestType);
    }

    public void reserveRoom(int aRoomNumber) {
        roomAt(aRoomNumber).reserve();
    }

    // accounting

    public int totalLosses() {
        return totalOfRooms(room -> room.lossUsingIfAbsentGuestType(prices, Floor.unknownGuestTypeBlock()));
    }

    public int totalProfits() {
        return totalOfRooms(room -> room.profitUsingIfAbsentGuestType(prices, Floor.unknownGuestTypeBlock()));
    }

    // totals

    public int totalRooms() {
        return rooms.size();
    }

    public int totalRoomsAvailable() {
        return totalRoomsThat(Room::isAvailable);
    }

    public int totalRoomsOccupied() {
        return totalRoomsThat(Room::isOccupied);
    }

    public int totalRoomsReserved() {
        return totalRoomsThat(Room::isReserved);
    }

    // private

    private Room roomAt(int aRoomNumber) {
        if (aRoomNumber < 1 || aRoomNumber > totalRooms()) {
            Floor.signalRoomNumberDoesNotExist();
        }

        return rooms.get(aRoomNumber - 1);
    }

    private int totalOfRooms(ToIntFunction<Room> aRoomAmount) {
        return rooms.stream().mapToInt(aRoomAmount).sum();
    }

    private int totalRoomsThat(Predicate<Room> aCondition) {
        return (int) rooms.stream().filter(aCondition).count();
    }

    private static Supplier<Integer> unknownGuestTypeBlock() {
        return () -> {
            Floor.signalUnknownGuestType();
            return null;
        };
    }

    // assertions (class side)

    public static void assertArePrices(Map<String, Integer> aPriceList) {
        if (aPriceList.isEmpty()) {
            Floor.signalNoPrices();
        }

        for (Integer price : aPriceList.values()) {
            Floor.assertIsPositiveIfFalse(price, () -> Floor.signalPriceMustBePositive());
        }
    }

    public static void assertIsIntegerIfFalse(Number aNumber, Runnable signalsBlock) {
        if (!(aNumber instanceof Integer)) {
            signalsBlock.run();
        }
    }

    public static void assertIsPositiveIfFalse(Number aNumber, Runnable signalsBlock) {
        if (!(aNumber.doubleValue() > 0)) {
            signalsBlock.run();
        }
    }

    // error messages (class side)

    public static String numberOfRoomsMustBeIntegerErrorDescription() {
        return "Number of rooms must be integer";
    }

    public static String numberOfRoomsMustBePositiveErrorDescription() {
        return "The number of rooms must be positive";
    }

    public static String priceMustBeIntegerErrorDescription() {
        return "Price must be integer";
    }

    public static String priceMustBePositiveErrorDescription() {
        return "Price must be positive";
    }

    public static String pricesListCannotBeEmptyErrorDescription() {
        return "Prices List must not be empty";
    }

    public static String roomNumberDoesNotExistErrorDescription() {
        return "Room number does not exist";
    }

    public static String unknownGuestTypeErrorDescription() {
        return "The guest type is unknown";
    }

    // exceptions (class side)

    public static void signalNoPrices() {
        throw new RuntimeException(Floor.pricesListCannotBeEmptyErrorDescription());
    }

    public static void signalNumberOfRoomsMustBeInteger() {
        throw new RuntimeException(Floor.numberOfRoomsMustBeIntegerErrorDescription());
    }

    public static void signalNumberOfRoomsMustBePositive() {
        throw new RuntimeException(Floor.numberOfRoomsMustBePositiveErrorDescription());
    }

    public static void signalPriceMustBeInteger() {
        throw new RuntimeException(Floor.priceMustBeIntegerErrorDescription());
    }

    public static void signalPriceMustBePositive() {
        throw new RuntimeException(Floor.priceMustBePositiveErrorDescription());
    }

    public static void signalRoomNumberDoesNotExist() {
        throw new RuntimeException(Floor.roomNumberDoesNotExistErrorDescription());
    }

    public static void signalUnknownGuestType() {
        throw new RuntimeException(Floor.unknownGuestTypeErrorDescription());
    }
}
